sdk/go/x/auto/errors.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2016-2020, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auto
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+type autoError struct {
+	stdout string
+	stderr string
+	code   int
+	err    error
+}
+
+func newAutoError(err error, stdout, stderr string, code int) autoError {
+	return autoError{
+		stdout,
+		stderr,
+		code,
+		err,
+	}
+}
+
+func (ae autoError) Error() string {
+	return errors.Wrap(
+		ae.err, fmt.Sprintf("code: %d\n, stdout: %s\n, stderr: %s\n", ae.code, ae.stdout, ae.stderr),
+	).Error()
+}
+
+// IsConcurrentUpdateError returns true if the error was a result of a conflicting update locking the stack.
+func IsConcurrentUpdateError(e error) bool {
+	ae, ok := e.(autoError)
+	if !ok {
+		return false
+	}
+
+	return strings.Contains(ae.stderr, "[409] Conflict: Another update is currently in progress.")
+}
+
+// IsCompilationError returns true if the program failed at the build/run step (only Typescript, Go, C#)
```

Comment:
I implemented `IsCompilationError` for mostly demonstrative purposes. The only error here in this package that I think is truly useful is `IsConflictingUpdateError`.

I need more real usage like @metral's to identify what cases to handle. He had good thoughts about having an AutoError that indicates if stack creation failed because one already exists: https://github.com/pulumi/pulumi-kubernetes-operator/pull/86/files#diff-e31cd52d2ed37a9f588a46e7ca7e52b6R551-R556

Reply:
With that said, `IsCompilationError`, and likely `UnexpectedEngineError` and `IsRuntimeError` can probably be removed.